src/deblock.rs

```diff
@@ -0,0 +1,439 @@
+    let mode_type = if mode >= NEARESTMV && mode != GLOBALMV && mode!= GLOBAL_GLOBALMV {1} else {0};
+    let idx = if pli == 0 { pass } else { pli+1 };
+    let block_delta = if fi.deblock.block_delta_multi {
+        block.deblock_deltas[ idx ] << fi.deblock.block_delta_shift
+    } else {
+        block.deblock_deltas[ 0 ] << fi.deblock.block_delta_shift
+    };
+
+    let level = clamp(block_delta + fi.deblock.levels[idx] as i8, 0, MAX_LOOP_FILTER as i8) as u8;
+    // if fi.seg_feaure_active {
+    // rav1e does not yet support segments or segment features
+    // }
+    if fi.deblock.deltas_enabled {
+        let l5 = level >> 5;
+        clamp (level as i32 + ((fi.deblock.ref_deltas[reference] as i32) << l5) +
+               if reference == INTRA_FRAME {
+                   0
+               } else {
+                   (fi.deblock.mode_deltas[mode_type] as i32) << l5
+               }, 0, MAX_LOOP_FILTER as i32) as u8
+    } else {
+        level
+    }
+}
+
+// Only ever called on a tx edge
+fn deblock_params(fi: &FrameInvariants, bc: &mut BlockContext, in_bo: &BlockOffset, p: &mut Plane, pli: usize, pass: usize, block_edge: bool, bd: usize) -> (usize, usize, u8, u16, u16, u16) {
+    let mut bo = in_bo.clone();
+    let w = p.cfg.width;
+    let h = p.cfg.height;
+    let xdec = p.cfg.xdec;
+    let ydec = p.cfg.ydec;
+    let po = bo.plane_offset(&p.cfg);
+        
+    // at or past edge of the frame?  Don't deblock, signal next loop.
+    if (po.x >= w) || (po.y >= h) {
+        return (0, 0, 0, 0, 0, 0)
+    }
+
+    // This little bit of weirdness is straight out of the spec;
+    // subsamped chroma uses odd mi row/col
+    bo.x |= xdec;
+    bo.y |= ydec;
+    let block = bc.at(&bo);
+    let block_adv = if pass == 0 { cmp::max(block.n4_w, 1<<xdec) } else { cmp::max(block.n4_h , 1<<ydec)};
+    let tx_adv = if pass == 0 { cmp::max(block.tx_w, 1<<xdec) } else { cmp::max(block.tx_h , 1<<ydec)};
+    
+    // upper/left edge skipped
+    if (pass == 0 && po.x == 0) || (pass == 1 && po.y == 0) {
+        return (tx_adv, block_adv, 0, 0, 0, 0)
+    }
+    // We already know we're not at the upper/left corner, so prev_block is in frame
+    let prev_block = bc.at(&bo.with_offset(if pass==0 { -(1 << xdec) } else { 0 },
+                                           if pass==0 { 0 } else { -(1 << ydec) }));
+
+
+    // filter application is conditional on skip and block edge
+    if !(block_edge || !block.skip || !prev_block.skip ||
+         block.ref_frames[0] <= INTRA_FRAME || prev_block.ref_frames[0] <= INTRA_FRAME) {
+        return (tx_adv, block_adv, 0, 0, 0, 0)
+    }
+
+    let mut level = deblock_level(fi, block, pli, pass) as u16;
+    if level == 0 { level = deblock_level(fi, prev_block, pli, pass) as u16; }
+    if level == 0 {
+        (tx_adv, block_adv, 0, 0, 0, 0)
+    } else {
+        
+        let tx_size = cmp::max(if pass == 0 { block.tx_w>>xdec } else { block.tx_h>>ydec } , 1);
```

Comment:
Same as above, if you like that style.

Reply:
yes, I should definitely do a readability pass, as well as add a bunch more commentary.  I got overexcited at it working :-)